### PR TITLE
NodeInfo 2.0

### PR DESCRIPTION
**Pull request by SuperTux88**
I created a new version 2.0 and changed the following things:

#15: Make `software.name` free text (I didn't get much feedback to the issue, so I hope it's OK.)
#16: Use protocol names instead of software names (@annando I removed `friendica` and `gnusocial` and then added `dfrn` and `ostatus`. `zot` was already there, so I only removed `redmatrix`. I hope that is correct.)

That changes make NodeInfo more flexible for new projects, since they can support existing protocols without the need of a new NodeInfo-version. There can also be a fork of an existing software, and they can announce their new software name, but still support the same protocols.

The `software.name` should only be used as informational metadata (like `User-Agent` in http). When deciding if something is compatible, you should use `protocols` instead.

### CODE REVIEW

**Comment by annando** [comment]
Could you add "pnut" (see https://pnut.io/) as an inbound and outbound service as well and remove "appnet" from it? Additionally have a look at the "metadata" section of https://pirati.ca/nodeinfo/1.0 

There I have added everything that was not covered by the existing protocol.

File: schemas/2.0/schema.json [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      "description": "Metadata about server software in use.",
+      "type": "object",
+      "additionalProperties": false,
+      "required": [
+        "name",
+        "version"
+      ],
+      "properties": {
+        "name": {
+          "description": "The canonical name of this server software.",
```
Comment (jhass) [inline schemas/2.0/schema.json]: Maybe remove "canonical" here and expand on the user agent like meaning.

Comment (SuperTux88) [inline schemas/2.0/schema.json]: Hmm, I think the description is still correct, because the software should still choose and use a "canonical name" here. But it's not limited in the spec anymore, that you can name your software only "diaspora", "frienidca", "hubzilla" or "redmatrix", so we don't need to update it every time there is a new software, like "social-relay" or "socialhome" (they are already used, and with this change they are now valid).

**Comment by SuperTux88** [comment]
@annando I removed "appnet" and added "pnut" as services. Then I also added "atom1.0" and "rss2.0" as protocols. I also added "twitter" to the inbound services.

What I don't understand: How is "smtp" used as outbound service? Can the user add an email-address to their contacts and send a post as mail to them? Then I think you should use "smtp" as outbound protocol, since the server sends the mail (and that is already there). Or does the user really set their own smtp-credentials in the settings and use that to send mails? Then I'll add it to the "services" list.

If I didn't miss anything, everything else should already be there.

**Comment by annando** [comment]
We can add contacts as SMTP contacts. Then the status messages will be send via SMTP to them. The own inbox will be pulled via IMAP and we use the "mail" command from PHP to send the mails.

File: schemas/2.0/schema.json (hunk 1 above)
Comment (jhass) [inline schemas/2.0/schema.json]: Then I would define some rules nonetheless, like all-lowercase, a-z0-9 only or so.

Comment (SuperTux88) [inline schemas/2.0/schema.json]: Sounds good, but I would also add `-` to keep "social-relay" a valid value.

**Comment by SuperTux88** [comment]
When you use SMTP directly on the server to send the mail, I think you should add "smtp" as outbound protocol and not as a service (and "smtp" as protocol is already in the schema). I think it is only a service when you use the users smtp-account to send the mail.

I don't understand the part with IMAP, does that mean you can read your mails with friendica? So you have "imap" as inbound service?

**Comment by annando** [comment]
Yeah, it is possible to use Friendica as an (restricted) mail reader. We are speaking IMAP to query inboxes, but are using the "mail" command at the moment (but eventually will have to use SMTP to use the user's mail server for spam protection reasons).

Since we rely on IMAP and SMTP servers I think it could count as "services" and should only count as "protocols" if the system provided an SMTP or IMAP server itself.

But on the other side: Querying an IMAP server is comparable to import an RSS feed. And outgoing traffic to an SMTP server is comparable to outgoing traffic to another server with protocol XYZ.

**Comment by SuperTux88** [comment]
I don't think that the protocol needs to be implemented inside friendica to count as protocol, when calling the "mail" function in PHP I think that counts still as protocol.

I would difference protocol and service with this rule:
* if the users can simply add a mail-address or rss-feed-url to their contacts (and the server does the sending or receiving), then it's a protocol.
* if the users need to connect their account with an external account (IMAP/SMTP), then it's a service.

So when you can read IMAP with friendica, I'll add "imap" as inbound service. As long as the server does the SMTP, I think you should announce "smtp" as outbound protocol. But I can add "smtp" also to outbound services, for the case that you need to switch to the users own smtp servers.

Maybe we should also add "xmpp" as protocol, for servers with enabled chat?

**Comment by SuperTux88** [comment]
Oh, I just saw that we also already have "smtp" as outbound service. :)

**Comment by annando** [comment]
XMPP should only count as protocol when the content was shared with it. Just enabling the chat doesn't count in my eyes.

Concerning SMTP and IMAP it sounds okay to add them to both - but it should be done with a detailed explanation.

File: schemas/2.0/schema.json [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        "outbound"
+      ],
+      "properties": {
+        "inbound": {
+          "description": "The protocols this server can receive traffic for.",
+          "type": "array",
+          "minItems": 1,
+          "items": {
+            "enum": [
+              "atom1.0",
```
Comment (denschub) [inline schemas/2.0/schema.json]: I feel really weird about having `atom` and `rss` considered a protocol for inbound traffic. After all, these are just formats for representing syndication elements and I can't really consider that a "communication protocol". Both are formats for specifying how items could look like, but I am having a really hard time considering that a communication inside a social network.

Nodeinfo is designed "to expose metadata about an installation of a distributed social network" and not "to expose metadata about any possible way my software can work", and I really can't consider both RSS and Atom as a protocol for social communication. Any other opinions?

Comment (annando) [inline schemas/2.0/schema.json]: I'm using it to show that Friendica can import feeds based upon this protocol.

Comment (denschub) [inline schemas/2.0/schema.json]: Yes, nice feature, but how is this related to the communication between nodes?

Comment (annando) [inline schemas/2.0/schema.json]: It is a protocol that is used for incoming content. I do see your point, since it isn't a bidirectional way. But when the description was meant only for bidirectional communication then there needn't to be a separation between inbound and outbound.

Comment (denschub) [inline schemas/2.0/schema.json]: So, I just started this discussion with @SuperTux88 on IRC and I felt like we have to do it elsewhere, which is why I moved it back to GitHub. As it turns out, this is more of a "what NodeInfo should be" discussion than anything else. Right now, it's described as

> The two key goals are being able to get better insights into the user base of distributed social networking and the ability to build tools that allow users to choose the best fitting software and server for their needs.

Which then, obviously, would be fine. However, everyone apparently starts using NodeInfo as a way to determine how you can communicate with a given node. Which is pretty cool, but that changes the meaning of NodeInfo qutie a bit, doesn't it?

If we consider the list of protocols in a nodes NodeInfo to actually *mean* something in regards of the communication capabilities, we have to be careful. So basically, there are two possible ways of dealing with NodeInfo:

1. NodeInfo can be used as a way of exposing what a node can or can't do, but it doesn't really mean anything. If a Node says it can accept "diaspora", "friendica" and "statusnet", it's good to know, but we don't care. To determinate the best way a node should be communicated with is still the canonical name, so if the Node is a diaspora\* node, it's best using the diaspora protocol.
2. NodeInfo exposes meaningful information about what a node can actually communicate with. If a node wants to send something to another node, it checks its targets NodeInfo and picks the protocol that suits the needs for the current message. The sending node does not care about what software is actually running, since the list of protocols is authoritative and binding, as a sending node knows exactly what it can send to the other node.

If we go with 1, adding stuff like rss is probably not a bad idea. It adds some informational value, to maybe inform users, but there is not much point in basing any decision regards the actual node to node-information on this information. So using the list of protocols inside implementations should be a no-no then.

Version 2 would probably be the better way to achieve global interoperability. If a node sees a software adds "diaspora" to the list, this node can be certain it can deliver message to the target using the diaspora protocol, and the receiving end will be able to parse it. This way, we completely could ignore the canonical name and only decide which protocol to use by the list of supported inbound protocols.

If way 2 is what we want, that is using NodeInfo as a authorative way of deciding which language to talk to another node, then adding "rss" feels wrong. After all, what's the "social network communcation" component of RSS? Clearly, the RSS spec is clear about it's uses and if not even the author is a required field in RSS, what is the purpose of using that for communication between social network nodes?

Comment (SuperTux88) [inline schemas/2.0/schema.json]: @denschub Your "version 1" is what we have now with 1.0. That was only meant to be used from something like the-federation.info or podupti.me. For federation this is useless, we can't use the canonical name, because we don't know names from the future (so if there is a new software supporting the diaspora protocol, it wouldn't be able to communicate, because we don't know the name, and we need to update every pod to be compatible with it). And you can't see from the name, if a software supports the diaspora protocol, because sometimes it needs to be activated as plugin.

That's why I created this PR with 2.0. The main change is opening `software.name`, so there can be a new software using existing versions of NodeInfo and supporting existing protocols, without changing existing servers to be compatible with it.

I planned to use `protocols.inbound` for deciding if a server is compatible with the "diaspora" protocol. It doesn't matter if there is "rss" in the list, because I don't care, because "diaspora" would be the much better protocol (compared to "rss") if available.

But if we only want to have protocols for communication **between** nodes in the list, then "rss" shouldn't be in the protocols list. But then @annando is right, we can simple create one list and drop the `inbound`/`outbound` separation. A server is then simply able to communicate using this protocols.

So I think "rss" should go to the `services` list, and services is then a list of "features" available on a server to import/export stuff to somewhere else. It doesn't matter if you need an external account for a service. And we can keep the `inbound`/`outbound` separation for `services`, because it is probably most times only in one direction.

Comment (annando) [inline schemas/2.0/schema.json]: I do agree with @SuperTux88. Putting RSS to ```services``` sounds like a reasonable idea.

```protocols``` then would only be used for determine what servers are able to speak bidirectionally (so no separation anymore between ```inbound``` and ```outbound``` anymore). Then it is totally logical that SMTP and IMAP are ```services``` as well - but not protocols.

BTW: Concerning the ```services```: There we need ```gnusocial``` (and not ```ostatus```) since Friendica can use GNU Social's API to communicate with GNU Social accounts.

Comment (denschub) [inline schemas/2.0/schema.json]: 👍 for adding RSS and Atom to `services` instead of `protocols`.

However, I'd still keep the `inbound`/`outbound` protocol seperation since I can imagine a situation where one application for example is able to receive `diaspora`, `friendica` and `gnusocial`, but will only ever send `gnusocial` outgoing.

Comment (SuperTux88) [inline schemas/2.0/schema.json]: At least for the diaspora-protocol, you can't just support receiving, or you would ignore half of the protocol. Relayables needs sending, so you either support it or you don't support it, you can't support it half, it would be broken then. A special case for the diaspora-protocol is the social-relay, where it only proxies messages, but even there [it is added to both](https://relay.iliketoast.net/nodeinfo/1.0), so if you want to separate it, we would need to split it with `inbount`, `outbound` and `proxy` (or we can split up each protocol and write how many parts the server supports), and that doesn't make much sense.

I don't know about other protocols (if you can support only receiving, without sending). Would you add "rss" again to `protocols`, if you want protocols that work in only one direction?

The only thing that is interesting is "can I send messages to that server using protocol x?", so I really think we can simplify `protocols` here to "this server supports this protocols", so we know we can send messages to it.

Comment (denschub) [inline schemas/2.0/schema.json]: Hm, okay. Agreed. Let's kill `inbound`/`outbound`.

**Comment by SuperTux88** [comment]
I updated the PR:
* add regex for `software.name`
* kill `inbound`/`outbound` for `protocols`
* move "atom1.0" and "rss2.0" to `services`
* add "imap" as inbound service

Some last questions:
* Should we remove "smtp" from `protocols`? (It was there in 1.0, but I don't know any "distributed social network" using SMTP to communicate between nodes.)
* Should we add "xmpp" to `services`? (In 1.0 we used `metadata.xmppChat` as unofficial property for the chat. At https://podupti.me/ it is listed as "service", but it isn't a normal service, and it also doesn't make much sense to split "xmpp" to `inbound`/`outbound`.)

@annando Can you check the `services` lists again, if everything is there that you need?

**Comment by Gargron** [comment]
Hey, I am not sure what particular info would be useful to exchange between servers through this protocol but having a standardized way of identifying server software is immensely useful, I am all for it. Will try to integrate this protocol into Mastodon soon!

**Comment by annando** [comment]
You should add "activitypub" to the lists of the protocols as well: https://www.w3.org/TR/activitypub/

Additionally I don't think that "mediagoblin" could be called a protocol. They use pump.io in the past, I think. And it seems they are now heading towards activitypub.

"xmpp" could be part of the protocols as well - see for example "movim".

Concerning "smtp": I haven't heard of a smtp based network by now. But this sounds imaginable, so I would keep it.

"xmpp" in the services would only make sense when xmpp could be used to transfer messages from the system, I think. This is a huge difference to simply having a web chat installed on a system.

**Comment by strugee** [comment]
@annando MediaGoblin has federation support based on the pump.io protocol sitting in an (unmerged) branch; the plan is to convert that to use ActivityPub and then merge it.

**Comment by SuperTux88** [comment]
@annando I updated the PR:
* Added "pop3" to inbound `services`
* Added "activitypub" to `protocols`
* Removed "mediagoblin" from `protocols`

> "xmpp" could be part of the protocols as well - see for example "movim".

I don't know if "xmpp" is specific enough as protocol between nodes. "xmpp" is for me more like "http", so I think we need to be more specific here. I think "buddycloud" (which is as separate protocol already in the list) also uses "xmpp", but I don't know if that is compatible with "movim"? So we should probably simply add "movim" as protocol? (or how is the protocol named?)

Same applies somehow to "smtp" as protocol, I don't think that it is specific enough for communication between nodes in social networks. So we should probably remove "smtp" from `protocols` until there is a social network using it, and then look how we would add it again?

> "xmpp" in the services would only make sense when xmpp could be used to transfer messages from the system, I think. This is a huge difference to simply having a web chat installed on a system.

So `services` should only contains stuff, that can handle posts? And since the xmpp-chat is a separated system (only sharing diaspora-ID and aspects), it wouldn't be a service. I noticed that "xmpp" is included in `services.outbound` 1.0. We can keep that, but it doesn't have anything to do with the chat, it would mean that you can send a post (or something else) to contacts via xmpp?

**Comment by annando** [comment]
Movim seems to work with regular XMPP clients: https://en.wikipedia.org/wiki/Movim#Concept

For everything else I totally agree with you.

**Comment by SuperTux88** [comment]
> Movim seems to work with regular XMPP clients: https://en.wikipedia.org/wiki/Movim#Concept

Hmm, that's about clients, but I think it's probably regular xmpp between servers too then? So I added "xmpp" as protocol for movim.

> For everything else I totally agree with you.

So I removed "smtp" from protocols. We can maybe add this later again, when someone is using it and we know the specs. Since every current software would need to add the protocol in the code too, it wont be a problem to add it in a later version of NodeInfo.

We could add some more stuff to the `services.inbound` list, but I don't want to add simply everything since nobody is using it now, we can also add it in a later version.

@annando @jhass @denschub So I think it's time for another review and look if it contains everything that we need now :)

**Comment by Flaburgan** [comment]
Pinging @edhelas for movim. But that makes me wondering, do we want to add protocols for projects that are not using nodeInfo, and never said they are going to? Isn't that making the things heavier and more complicated?

**Comment by annando** [comment]
I don't see a problem there. Making nodeinfo a standard across different distributed networks sounds like a very great idea for me.

**Comment by SuperTux88** [comment]
So are there any open points/questions here? Or can that be merged/released?

> But that makes me wondering, do we want to add protocols for projects that are not using nodeInfo, and never said they are going to?

Projects that are using NodeInfo can still support that protocols (and then add it to their NodeInfo).

**Comment by annando** [comment]
Just one single question: Should a system always offer only one version (the latest) or should it offer 1.0 as well as 1.1 and 2.0?

**Comment by SuperTux88** [comment]
A system should always offer all versions they support and the client should then request the latest version they understand. So you should still offer 1.0, because old pods only know about 1.0 and can't parse 2.0.

See http://nodeinfo.diaspora.software/protocol.html: "A client should follow the link matching the highest schema version it supports."

So for now that is 1.0 and after this is merged/released 1.0 and 2.0. I think 1.1 was never officially released?

**Comment by jaywink** [comment]
At the risk of poking my head again to places I shouldn't, but since I was pinged by @Flaburgan , just giving a total IMHO opinion that still NodeInfo will be too restrictive for wide usage. IMHO, `protocols` and `services` cannot be locked strings in the spec. This is just shooting oneself in the foot as at some point someone will want to use something else and suddenly the NodeInfo document cannot be generated. This is especially true for services. You can't *possibly* keep the spec up to date with all possible platforms out there. It's already missing some very popular services.

To be usable NodeInfo should IMHO allow any values to be given. It is *the reader* who is responsible to understand or not understand the values they find. The spec should give recommendations but it shouldn't stop you generating a document with your fancy new protocol or a fancy new service you've integrated to.

Until this happens, IMHO NodeInfo is not usable outside the Diaspora (protocol) ecosystem.

**Comment by SuperTux88** [comment]
We had this topic already multiple times (https://github.com/jhass/nodeinfo/pull/12 and https://github.com/jhass/nodeinfo/issues/15#issuecomment-250263612) ...

If you have a new protocol or a new service, you can create an issue/pr here and it can be added and released in a short time (probably a few hours to a few days). Every other software doesn't need to change anything, until they support the new protocol or service. And with this the strings used for protocols/services are defined in the schema and standardized. And I don't think that creating an issue/pr here is too complicated or restrictive. After that you can generate the document with your fancy new protocol/service.

> It's already missing some very popular services.

I added all services currently supported by friendica, which weren't in the previous version. If you know any other service supported by anyone, you can list them here and I'll add them.

> It is the reader who is responsible to understand or not understand the values they find.

But the reader needs to know which values to expect. If everyone just adds stuff and everyone uses it's own spelling of a protocol/service, that doesn't help anyone, it only adds chaos.

> Until this happens, IMHO NodeInfo is not usable outside the Diaspora (protocol) ecosystem.

This new version is perfectly usable for everyone, they can announce their software supporting every existing protocol and they are immediately compatible with the other software supporting the same protocol. If they have a new protocol they can create an issue/pr here and it will be added. But since other existing software don't know the new protocol, they can update to the new NodeInfo version including the new protocol when they add support for the new protocol. So there is no problem for others to use NodeInfo and adding their protocol to the list.

**Comment by annando** [comment]
@jaywink we are currently addressing other projects so that they use this protocol as well. Maybe @cwebber is interested in mentioning this in the ActivityPub documentation as well.

**Comment by Flaburgan** [comment]
>But the reader needs to know which values to expect.

I agree about that. Is the value `diaspora`, or `diaspora*`, or `Diaspora` or... ? It means we have to list the possible values somewhere anyway, so inside the project is probably the best place.

**Comment by annando** [comment]
@Flaburgan it's even worse for Friendica, Hubzilla and GNU Social. The protocol there isn't called like the software - but people could mix that.

**Comment by SuperTux88** [comment]
That's why `protocols` is stricter than `software`, because you need `protocols` to decide with which protocol you talk to another server, so you need to know what's in this list. But `software.name` and `software.version` shouldn't be used for decisions, it's more for statistics or debugging (with which software and which version did you have that problem), that's why that is opened now, because that can change when you create a new software or fork (but protocols still stays the same).

**Comment by jhass** [comment]
Let's do this!

**Comment by jaywink** [comment]
> Maybe @cwebber is interested in mentioning this in the ActivityPub documentation as well.

I certainly *hope not*. This iteration of NodeInfo is an improvement for sure in representing nodes in the "Diaspora protocol and friends" network - but that is all it is good for. If you design something and lock down the details, that something can only be used for what the designers could imagine for.

There is **already** one important decentralized social protocol missing from the enforced list that is being used. There are **many** popular services already missing from the services list. Having to do iteration after iteration and output 10x different NodeInfo documents forever with a different schema is hardly a very flexible system of solving a simple task - that simple task answering the question: "what does my node do and what is the usage".

Besides, when talking about ActivityPub, it would be hard to recommend something that requires using a `.well-known` url, which ActivityPub doesn't use at all. I guess an Actor object could refer to a NodeInfo URL, but that wouldn't support all these iterations that need to be supported, already counting to 3 versions :)

**Comment by annando** [comment]
Could you tell what exactly is missing?

**Comment by jaywink** [comment]
> Could you tell what exactly is missing?

No, because I don't claim to know all the possible protocols or services that might exist. But I do know one protocol that I'm using myself (not as author but user). It's actually used by three different software stacks AFAIK. Maybe you can start guessing ;) Or not. I don't particularly want to support this model so whatever ;)
